### PR TITLE
List View: Allow Escape key to deselect blocks if blocks are selected

### LINKED ISSUE #48462: List view: Impossible to deselect all blocks using the keyboard
### Description

It is impossible to deselect all blocks using the keyboard only. If you select some blocks, there is no way to remove your selection. I would expect this to not be the case.

### Step-by-step reproduction instructions

1. Open the editor and insert 5 blocks.
2. Open the list view.
3. Navigate to the 2nd block in the list view.
4. Press shift+down arrow.
5. Press shift+up arrow.
6. Notice how there are still blocks selected indicated by the live region at the bottom of the page.
7. There should be a way to deselect all blocks without selecting the 1st block causing the selection to go to 2 again.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Browser: Firefox (110.0)
WordPress: 6.2-alpha-54678
Gutenberg: trunk

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Fixes #48462 

## Related issues and discussions:

* https://github.com/WordPress/gutenberg/issues/36535
* https://github.com/WordPress/gutenberg/issues/47172

This is an attempt to solve the issue of users being unable to deselect blocks from within the list view. The idea proposed is to allow the Escape key to deselect all blocks if there is a block selection, and skip closing the list view. If there are no blocks selected, then the Escape key will close the list view.

## Why?

As reported in #48462, it would be good for users to be able to deselect blocks from within the list view via keyboard, and it is not currently possible to do so. Because folks can use the list view keyboard shortcut to open and close it, the idea in this PR is that it would be okay for Escape to be prioritised for deselecting blocks rather than closing the list view.

For reference, the keyboard shortcut for closing the list view is alt+shift+o on Windows, and control+option+o on Mac.

## How?

* In the list view if there is a block selection when the user presses the Escape key, then clear the block selection instead of closing the list view. Otherwise, proceed to closing the list view.

## Testing Instructions

1. In a post with multiple blocks, try shift selecting a couple of them block from the list view.
2. Press Escape and the block should be deselected.
3. Press Escape again and the list view should be closed.
4. Open the list view again, and tab to the items within the list view without selecting any of the blocks. Press Escape and the list view should close.

### Testing Instructions for Keyboard

1. In a post with multiple blocks, open the list view via the keyboard shortcut (alt+shift+o on Windows, and control+option+o on Mac).
2. Use shift + up/down keys to select multiple blocks.
3. Press Escape and the blocks should be deselected.
4. Press Escape again and the list view should be closed.

## Screenshots or screencast 

In the below screengrab, the user shift-clicks multiple blocks within the list view, then presses Escape to clear the selection. When there are no blocks selected and the user presses Escape, the list view is closed.

https://user-images.githubusercontent.com/14988353/222635879-f751e946-9cde-403e-9383-ea45126f09ec.mp4

### CODE REVIEW

**Comment by andrewserong** [comment]
@alexstine just for visibility, here's the draft I'm working on to try out deselecting blocks within the list view via using the Escape key. I wasn't sure, but for this one, will we need to add a `speak()` call to announce that the blocks have been deselected?

**Review by alexstine — COMMENTED** [review]
@andrewserong This mostly looks good but yeah, we'll need a `speak( msg, 'assertive' );` on this. Maybe model it after the announcement in the current selection hook in the list-view.

```
speak( 'All blocks deselected', 'assertive' );
```

Is it okay to access `blockEditorStore` outside of block-editor package?

**Comment by andrewserong** [comment]
Thanks for the feedback @alexstine! I've added the assertive `speak()` call to each of these callbacks. I noticed that the selection hook in the list view doesn't currently use `assertive`, so I wasn't sure which approach is better. Happy to keep tweaking as needed.

> Is it okay to access blockEditorStore outside of block-editor package?

Yes, in the case of the edit-post and edit-site packages, the block editor store appears to be a safe dependency, and it's used quite a bit for block selection actions and selectors.

**Comment by andrewserong** [comment]
Just including @WordPress/gutenberg-design as a ping since I know a few designers have helped out with UX flows of the List View in the past, so might have feedback on the approach here, too 🙂

**Comment by alexstine** [comment]
@andrewserong I am working on changing that in my ARIA refactor, #48461. I will give this PR a test shortly.

Thanks.

**Comment by andrewserong** [comment]
> I am working on changing that in my ARIA refactor, https://github.com/WordPress/gutenberg/pull/48461.

Ah, yes, of course I forgot about that. Thank you for the reminder!

**Comment by jasmussen** [comment]
Hey, thanks for taking a stab at this, this is really important to get right.

This issue seems to be working for the list view, but I wonder, can we find a more basic solution that works for blocks in the canvas too? In #47172 I had an idea to allow esc twice to work for this. Would that thread the needle?

**Comment by alexstine** [comment]
@jasmussen I think pressing escape twice is a little excessive for this. Besides, most dialogs/pop overs in Gutenberg close with a single press of escape which can sometimes be a very destructive action.

**Review by apeatling — COMMENTED** [review]
I realize there's some conversation about direction, just noting that I tested this as it currently stands and it all worked as expected.

**Comment by andrewserong** [comment]
Thanks for the feedback @jasmussen! Totally agree that this is an important one to get right, and to try think about the implications holistically. 

In the case of the list view, the behaviour in this PR only applies to when focus is within the list items, where a user is interacting with and adjusting selection of the blocks. For the Escape key to be able to trigger a deselection, we'll probably always need to have some code to handle it here, so that we can skip closing the list view, as Alex mentions.

> can we find a more basic solution that works for blocks in the canvas too? In https://github.com/WordPress/gutenberg/issues/47172 I had an idea to allow esc twice to work for this. Would that thread the needle?

For the editor in general, what if we thought about it the other way, and had logic that says that if there is a multiple block selection in place within the editor canvas, then the first Escape press clears the selection altogether, and then the second press switches to the Select mode? This is a sample size of one, but personally I'd expect deselection to happen first before a mode switch.

Whichever direction we go with, though, I agree that the behaviour in the List View should be consistent with expected behaviour elsewhere in the editor. I think this PR will likely still be a good shape for it within the List View, but I'm very happy for us to slow down a little on this PR to think it through — also happy to try other ideas if there are others in mind!

**Comment by jasmussen** [comment]
> I think pressing escape twice is a little excessive for this. Besides, most dialogs/pop overs in Gutenberg close with a single press of escape which can sometimes be a very destructive action.

Thanks for diving in. I tend to agree, and I suggested it mainly because I so wanted a way to deselect all blocks in the canvas using Esc, that I was willing to entertain the double-esc idea. 

I quite like Andrew's suggestion above, which is slightly simplified in [this comment](https://github.com/WordPress/gutenberg/issues/47172#issuecomment-1457511464). Essentially that if you have blocks selected, it deselects them. Select mode you only enter if you don't have any blocks selected. (Or have 1 block selected? — I guess that's the main inconsistency to figure out.) What do you think about that behavior?

**Comment by alexstine** [comment]
@jasmussen Agreed. We should not be allowing the user to enter select mode if blocks are selected or closing the list view if blocks are selected. I think the double escape will confuse users if nothing happens on the first press though. Besides, what if something like this happens?
1. Open list view.
2. Select some blocks.
3. Press escape.
4. Explore the current selection just using the arrow keys.
5. Do I still need to press escape twice? Does it reset after a certain period? For example, I must press escape twice within 3 seconds?
6. Press escape twice.
7. All blocks deselected.
8. Press escape once, list view closes.
I just cannot get behind the UX for that. It seems confusing and it introduces conditionals that are not really discoverable by users until they try it and figure it out.
As far as deselecting in the canvas goes, I think we'll have our own set of A11Y challenges there. That section of the editor does not have `role="application"` for screen readers so I think we'll see all kinds of A11Y issues with Windows screen readers. My focus for now is to get at least one area working and maybe it can be the framework for other parts of the editor.
Hopefully my thoughts make sense.
Thanks.

**Comment by andrewserong** [comment]
I realised that I forgot to include the widgets editor, too, which also has its own sidebar logic. In https://github.com/WordPress/gutenberg/pull/48708/commits/1d4c96f66e7168474b12e224e7a0a060def592a3 I've included that one, too.

Separately, I've had a go over in https://github.com/WordPress/gutenberg/pull/48904 of trying out allowing Escape in the editor canvas to clear the block selection. At the very least, it highlighted that if we want to allow block deselection within the list view, this PR will still be required, as we need the logic to fire before the Escape key handlers in the list views that are responsible for closing the list view. So, whichever direction we think will work best for the editor canvas, I _think_ that means this PR should still be viable. Or at the very least, is in a good place for review now 🙂

Let me know if anyone would like me to try out any other ideas not already covered in this PR and over in #48904 (that one just deals with the editor canvas, this one just with the list views).

Thanks again for all the discussion here!

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js
Comment (talldan) [inline packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Something I noticed is that if you have focus on the List View tab button or the close button for the sidebar and press escape, it still deselects all blocks. 

I wondered if it perhaps only makes sense to deselect blocks when focus is within the ListView component itself (i.e. on a block in List View). What do you think @andrewserong  @alexstine?

Comment (andrewserong) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Oh, great point. Yes, I think it makes more sense to only deselect while focus is _within_ the list view (i.e. on a block as you mention). I'll be AFK from tomorrow through Monday, so can have a go at updating the logic on Tuesday if I run out of time today. Keen to hear what Alex thinks, though.

**Comment by andrewserong** [comment]
Thanks for the suggestions @talldan of reducing the scope of where the deselection occurs, so that it's only when focused within the list view content area. I've tried that out in https://github.com/WordPress/gutenberg/pull/48708/commits/f3b91a0f0c68bbac8e024af3e5c1838389ce7fd4 — now if the focus is on the tabs above the list view, or on the close button, then the deselection will not occur, and the list view itself will close.

It's an easy revert, though, if Alex or anyone else prefers the handling to apply to the entire list view sidebar, rather than the list view panel content area.

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js
Comment (alexstine) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Agreed. 👍

**Comment by alexstine** [comment]
@andrewserong I found one bug. To replicate:
1. Open a post or page and have at least 5 blocks.
2. Select the last block, number 5.
3. Open the list view and select the first 2 blocks.
4. Press escape.
5. Notice how all blocks are deselected. This is fine.
6. Notice how the 5th block is not selected in the canvas anymore. I do not think this was supposed to happen. Escape should clear the selection, not the currently editing block.

Thanks.

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js
Comment (tellthemachines) [inline packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: I'm torn between asking for a "disable reason" comment here (in general it's good to add one when we disable a lint rule) and wondering if we should just get rid of this rule altogether, given that we're disabling it in so many places already 😅

**Review by tellthemachines — COMMENTED** [review]
Code looks good! 

In trying to reproduce the bug Alex found on macOS Chrome and Safari I'm finding that when I press escape after selecting the top two blocks in the list, focus always goes to the topmost block. Would the expectation be for focus to stay on the fifth block instead?

Another thing I'm noticing is when typing in the canvas with the list view open, and then pressing ctrl+opt+o to move focus to the list view, focus doesn't always land on the block I was editing in the canvas, but often a couple of blocks above it. I can also reproduce this on trunk so it's not related to this PR, but thought I'd mention it as it might be relevant to the bug Alex found.

**Comment by andrewserong** [comment]
Thanks for the feedback @alexstine and @tellthemachines!

> I found one bug. To replicate:

If I'm following along, the expectation is that if no blocks are selected when closing out the list view, we return focus to the previously selected block when the list view was initially opened? If so, I've had a go at implementing this in [this commit](https://github.com/WordPress/gutenberg/pull/48708/commits/b68fcced83120ac7c7f3406506124001588298e7). It's only working in the post editor for now, but if it works okay, then I can roll it out to the site and widget editors.

I'm a little torn on the behaviour — on balance, I think it makes switching between the editor canvas and list view and back feel a little more intuitive when clearing out the block selection, as it helps avoid the user being stranded at the top of the document. On the other hand, we can't quite know what the user has done since opening up the list view, so depending on how long it's been since they opened it, we might be returning the user to somewhere unexpected. Especially if they don't care about what they were working on previously to opening the list view.

Very happy for any feedback on the latest change, and also happy to roll back the change if it doesn't feel right. We could always look at the behaviour when closing the list view in a separate PR if it proves tricky to get right.

> In trying to reproduce the bug Alex found on macOS Chrome and Safari I'm finding that when I press escape after selecting the top two blocks in the list, focus always goes to the topmost block.

When you press Escape to deselect blocks within the list view, focus should still be on the last focused block within the list view. So if you press Shift+UP to select the top two blocks, then after pressing Escape I'd expect focus in the list view to be on the top-most block. If you pressed Shift+DOWN to select the top two blocks, then after pressing Escape, I'd expect focus in the list view to be on the second block. Is that what's happening?

> Another thing I'm noticing is when typing in the canvas with the list view open, and then pressing ctrl+opt+o to move focus to the list view, focus doesn't always land on the block I was editing in the canvas, but often a couple of blocks above it.

This sounds like it might be because of the roving tab index logic in the TreeGrid component that keeps track of the last element that was focused within the list view ([the code is here](https://github.com/WordPress/gutenberg/blob/464426a5d439143c029fbd9879d893e2ad28ac13/packages/components/src/tree-grid/roving-tab-index-item.tsx#L20)). So, when focus is returned to the list view after pressing ctrl+opt+o, it'll return you to the last element you were working on within the list view, rather than the currently selected block. My understanding is that this behaviour was intentional?

---

Thanks again for all the feedback. Happy to continue this exploration with trying to return the user to the desired place when the list view is closed, or split it out to a follow up PR, whichever feels most practical 🙂

**Comment by alexstine** [comment]
> This sounds like it might be because of the roving tab index logic in the TreeGrid component that keeps track of the last element that was focused within the list view (the code is here). So, when focus is returned to the list view after pressing ctrl+opt+o, it'll return you to the last element you were working on within the list view, rather than the currently selected block. My understanding is that this behaviour was intentional?

This should have been fixed in my PR: https://github.com/WordPress/gutenberg/pull/48339

I will add this back to my review queue and check it out. Thanks @andrewserong .

**Comment by andrewserong** [comment]
> This should have been fixed in my PR: https://github.com/WordPress/gutenberg/pull/48339

Oh, I wonder if it's that a block that was focused but isn't selected is still receiving the roving tab index? So we might have a stray block in the list view with `tabIndex=0` in addition to the currently selected block 🤔

> I will add this back to my review queue and check it out.

Thanks Alex, appreciate it! No rush for a re-review if you're busy, I have plenty of other things to chip away at, too.

**Comment by kevin940726** [comment]
Do you think it's worth adding some integration/e2e tests? I'd be happy to help review them!

**Comment by alexstine** [comment]
I think we're trying to get this one in a good state first. Whole test suite around the list view likely needs revisiting at some point. Do not want regressions with all the hard work.

**Comment by andrewserong** [comment]
Thanks for the offer, Kai! Just echoing Alex's statement — once this PR is in good shape from a usability angle, let's then add in some e2e test coverage 👍

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js
Comment (kevin940726) [inline packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: What do you think if we try implementing this logic inside the shared `PrivateListView` component in `block-editor`? An immediate benefit is that we don't have to implement this for every editor we have 😆 . It might be possible to do this inside `use-block-selection`, if that makes sense.

Comment (draganescu) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Yes to what @kevin940726 said :)

Comment (andrewserong) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Oh, good idea! That'd be much neater if we can consolidate it as part of the list view component. The only thing is that we'd probably still need to have the secondary sidebar component be aware of when it shouldn't close the sidebar based on an Escape key press, which means the logic might be sort of split a little across the components. If we don't mind that, then I think it's a better option 👍

Comment (kevin940726) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Could we do `event.stopPropagation()` and/or `event.preventDefault()` to solve that?

Comment (andrewserong) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Good thinking 👍 It might depend on which way the propagation goes, so I think I might have had the wrong idea in mind that I needed to handle things in the sidebar. I'll have a play around with it next week and will report back how it goes!

**Comment by andrewserong** [comment]
Update: I've just given this a rebase and moved the logic over to the ListView block, as it's a much simpler way to handle it and avoids the duplication across each instance of the sidebar (thanks Kai!).

It currently clears out the whole block selection when a user has multiple blocks selected, and there's still the outstanding issue of what does this mean for focus when a user closes the list view, as no block will be selected, so it'd be easy to feel lost at that stage. One option could be for us to keep track of which block was selected when we initially open the list view, and then attempt to drop the user back to that block once the list view is closed — I'm not sure how stable that will feel in practice, but happy to hack around with it.

Also still to do: updating the e2e tests, but I might leave that to when we feel confident about the focus behaviour.

As always, happy for any feedback — at this stage, I think it's unlikely this particular PR will make it in time for 6.3, but you never know 🙂

**Comment by alexstine** [comment]
@andrewserong One thing I've been thinking a lot about is having a constant focus point in Gutenberg for when we have no idea how or where to return focus to. As a fallback, could you maybe return focus to the title field? That would make sense to me if no blocks are currently selected.

Thoughts? Thanks.

**Comment by andrewserong** [comment]
Good question @alexstine, the title field could work. Another potential option for the list view, could be the List View / Document Outline toggle button in the header toolbar, if no blocks are selected, as that'd make it easy to open the list view again. So, the idea there might be to have the focus be directed to a place that's related to the thing that has just been closed. Or would that be a frustrating place for the user to end up in?

Happy to try out ideas, but since this PR likely won't make it in time for 6.3, I'll probably put it on the back burner for a little bit while helping out with backports and bug fixes.

**Comment by draganescu** [comment]
This works great with multiple blocks selected, but if there is only one block selected in my testing ESC still triggered navigation mode. Is this intended?

**Comment by jasmussen** [comment]
For the time being, yes, I think we can separately explore changing what triggers select mode if need be.

**Review by draganescu — APPROVED** [review]
This works great for me doing exactly what is was described - a much better experience than the current situation. I would - if possible - try to figure out if _if the focus is in the List View_  to close the list view on ESC when one block is selected in the list view.

**Comment by alexstine** [comment]
So, we've gotten ourselves in a real mess as I highlighted in a previous PR. Select is a term that happens under two situations that are very different.
1. Select as in one block is currently editable.
2. Select as in multiple blocks are selected by holding down shift and using the arrow keys.

I am not sure how we managed to get ourselves in such a terrible naming convention but here we are. What further complicates things is the modes, navigation vs. edit mode. Due to the screen reader work arounds, I think you can select blocks from navigation mode.

Escape should close the list view if one block is active but escape should deselect if one or more blocks are selected.

Thanks.

**Comment by alexstine** [comment]
@andrewserong I think the toggle button works fine. I just wasn't sure if you could place focus there or not. Thanks.

**Comment by andrewserong** [comment]
> I think the toggle button works fine. I just wasn't sure if you could place focus there or not.

I'm not 100% sure how easy it'll be, but good to know that the toggle button will work for us, happy to give that a try 👍

> This works great for me doing exactly what is was described

Thanks for reviewing @draganescu, much appreciated. I think this PR isn't quite right just yet, though. The two things that would be good to resolve before merging are:

* When only one block is selected, allow Escape to close the list view as before, as Alex flagged
* If no blocks are selected when the list view closes, move focus back to the list view button

In the rush to finish things off just before the feature freeze coinciding with the GB 16.1 RC, I didn't wind up having time to fix up this PR and get it in. It'll still be good to land once I've fixed the remaining issues, but I expect it'll now be a feature for WP 16.4.

**Comment by andrewserong** [comment]
Now that we're in the 6.4 cycle, I've given this a rebase and added a small addition to the existing keyboard shortcuts test for the list view. I think it's mostly working pretty well, but I haven't yet found a good way to always return explicit focus to the list view button when the list view is closed via the Escape key without any blocks selected.

In some ways, that might be behaviour for us to look into separately, because it's possible to have focus within the list view without a block selected and then press Escape on `trunk`. That said, this PR makes that situation much more frequent, so happy to keep digging if anyone feels like it's a blocker to this PR landing.

**Review by alexstine — COMMENTED** [review]
@andrewserong Just gave this another test. While the selection announcement is working great, pressing Escape no longer gives any indication that blocks have been deselected. Was that an oversight on my part or is there a bug?

**Comment by andrewserong** [comment]
Thank you for re-testing @alexstine! It turns out it was a bug, I had forgotten that I needed to add extra handling to the `updateBlockSelection` callback to ensure that the `speak` calls for deselection continue to fire if the escape key is pressed. I've pushed a fix. Prior to that, the announcement code would never fire because we would return early if the shift key was not pressed. I believe we previously added in the deselection announcement only for actions where users are adjusting a multi selection via the shift key.

I've retested using VoiceOver on a Mac and the deselection announcements are now happening for me when I press the Escape key. Please let me know if I've missed anything, though, and I'll dig in.

**Comment by alexstine** [comment]
@andrewserong This is looking good. If you promise to look into the loss of focus in a follow-up, I'm good to let this get merged now. To replicate:
1. Open list view.
2. Press Escape to deselect blocks.
3. Press Escape to close the list view.
4. Focus should land somewhere and currently it goes nowhere. It would be good to focus the list view trigger or some other constant point.

Focus loss is a big problem but one I think for this situation, can be addressed in another PR.

Thanks.

**Review by alexstine — APPROVED** [review]
👍

**Comment by andrewserong** [comment]
Thanks for reviewing and approving @alexstine! Yes, absolutely, I'm happy to continue looking into the focus issue. I've opened up a separate issue over in #54165 to capture the discussion so far, and will have a deeper look.

Thanks again for all your help with this one! I'll merge this in now.

**Comment by jasmussen** [comment]
This is such a good improvement. Thank you all for working on this.

**Pull request merged**
